### PR TITLE
Add "debug" tab in devmode to draw ImGui UI

### PR DESCRIPTION
**Pull request by pierricgimmig**
This is a fairly big change, the main idea is to have a dedicated tab in devmode for
rendering various debug information with ImGui. It can also be used to expose 
interactive parameters, colors, toggles, etc. very easily, often with a single line. 

image

Reviving multiple OpenGL windows support will be used for UI introspection for example, 
but also for having multiple viewports for the same capture. Some ImGui cleanup has 
been done, but more is needed, mainly in ImGuiOrbit.h/.cpp.

- Only initialize ImGui in devmode
- Remove useless GlPanel.h/.cpp
- Add Render callbacks to GlCanvas 
- ImGui calls need to happen inside a RenderImGui scope

I suggest looking at the change in bulk and *not* by individual commits.

NOTE: The help text doesn't have a background anymore. It was done using ImGui before, but we don't
create an ImGui context for the capture view anymore. I am working on adding a fix.

### CODE REVIEW

**Comment by karupayun** [comment]
I like it! Seems clearly better to have the debug tab instead of drawing it on Capture Window.

BTW, as it is a big change, I will try yo review it by today but maybe I will do it tomorrow. :)

File: OrbitGl/ImGuiOrbit.cpp [hunk 1]
```diff
@@ -728,14 +738,51 @@ void OutputWindow::Draw(const char* title, bool* p_opened, ImVec2* a_Size) {
...
+}
+
+namespace ImGuiOrbit {
+std::string PrettyVariableName(const char* name) {
+  size_t length = strlen(name);
 
-  static int i1 = 0;
-  if (ImGui::SliderInt("slider int", &i1, 10, 100)) {
-    GCurrentTimeGraph->SetFontSize(i1);
+  // remove "m_"
```
Comment (florian-kuebler) [inline OrbitGl/ImGuiOrbit.cpp:749]: This feels a bit surprising. At least I would expect a comment what is going on here.

File: OrbitGl/ImGuiOrbit.cpp [hunk 2]
```diff
@@ -728,14 +738,51 @@ void OutputWindow::Draw(const char* title, bool* p_opened, ImVec2* a_Size) {
...
   ImGui::TextUnformatted(Buf.begin());
 
-  static bool checked = true;
-  ImGui::Checkbox("blah", &checked);
+  ImGui::End();
+  ImGui::PopStyleVar();
+}
+
+namespace ImGuiOrbit {
+std::string PrettyVariableName(const char* name) {
```
Comment (florian-kuebler) [inline OrbitGl/ImGuiOrbit.cpp:746]: Are we sure we want such a method? If I could follow correctly so far, this is pretty printing variables, that we expose explicitly to the debug view.
This complete method looks very fragile. I'd rather suggest to either, just expose the original variable names (would be easier to find in an IDE anyways), or when exposing them, also add a description string.

File: OrbitGl/TextRenderer.cpp [hunk 3]
```diff
@@ -166,8 +166,16 @@ void TextRenderer::AddTextInternal(texture_font_t* font, const char* text, const
   float strWidth = 0.f;
   int minX = INT_MAX;
   int maxX = -INT_MAX;
+  constexpr GLuint indices[6] = {0, 1, 2, 0, 2, 3};
```
Comment (florian-kuebler) [inline OrbitGl/TextRenderer.cpp:169]: Nit: I am not sure, if we have any convention yet. Our clang-tidy settings, complains about c-style arrays and suggest to use std::array instead. I don't have strong opinion on this, but just wanted to mention this.

File: OrbitGl/TimeGraphLayout.cpp [hunk 4]
```diff
@@ -36,47 +36,41 @@ TimeGraphLayout::TimeGraphLayout() {
...
-  FLOAT_SLIDER(right_margin_);
-  FLOAT_SLIDER(m_SchedulerTrackOffset);
-  FLOAT_SLIDER_MIN_MAX(m_TrackTabWidth, 0, 1000.f);
-  FLOAT_SLIDER_MIN_MAX(m_TrackBottomMargin, 0, 20.f);
-  FLOAT_SLIDER_MIN_MAX(m_TrackTopMargin, 0, 20.f);
-  FLOAT_SLIDER(m_ToolbarIconHeight);
-  FLOAT_SLIDER_MIN_MAX(scale_, 0.05f, 20.f);
-  ImGui::Checkbox("DrawTrackBackground", &m_DrawTrackBackground);
-  ImGui::End();
+  IMGUI_FLOAT_SLIDER(m_TrackLabelOffsetX);
```
Comment (florian-kuebler) [inline OrbitGl/TimeGraphLayout.cpp:47]: Can you please use this opportunity to update the names here to our style guide?

**Review by florian-kuebler — COMMENTED** [review]
Thanks Pierric for this change. I left some comments. Mostly nits, and nothing that I would insist in addressing before merging. Let me know what you think of those.

**Comment by karupayun** [comment]
I got this crash when I clicked in Misc... Could it be?

Orbit: /home/conan/.conan/data/imgui/1.69/bincrafters/stable/build/77263a2e24ef1df522d09982ea4ecda1262bf522/source_subfolder/imgui.cpp:4495: void CheckStacksSize(ImGuiWindow*, bool): Assertion `*p_backup == current && "PushID/PopID or TreeNode/TreePop Mismatch!"' failed.

File: OrbitGl/App.cpp [hunk 5]
```diff
@@ -473,6 +492,15 @@ void OrbitApp::RegisterCaptureWindow(CaptureWindow* capture) {
   capture_window_ = capture;
 }
 
+void OrbitApp::RegisterDebugCanvas(GlCanvas* debug_canvas) {
+  CHECK(debug_canvas_ == nullptr);
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitGl/App.cpp:496]: done above.

File: OrbitGl/App.cpp [hunk 6]
```diff
@@ -473,6 +492,15 @@ void OrbitApp::RegisterCaptureWindow(CaptureWindow* capture) {
...
+  constexpr uint32_t kDefaultFontSize = 14;
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitGl/App.cpp:497]: Yes, I added a GetInitialFontSize method to GlCanvas, cleaner now, thanks.

File: OrbitGl/CaptureWindow.cpp [hunk 7]
```diff
@@ -639,57 +626,47 @@ void CaptureWindow::RenderUI() {
...
-  Orbit_ImGui_NewFrame(this);
-
-#define VAR_TO_STR(var) VariableToString(#var, var)
-
-  if (draw_stats_) {
-    ImGui::ShowDemoWindow();
-    if (time_graph_.GetLayout().DrawProperties()) {
-      NeedsUpdate();
+void CaptureWindow::RenderImGui() {
+  if (ImGui::BeginTabBar("DebugTabBar", ImGuiTabBarFlags_None)) {
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitGl/CaptureWindow.cpp:632]: Normally yes, but this helps understand the hierarchy of ui elements, I prefer to keep it.

File: OrbitGl/CaptureWindow.cpp [hunk 8]
```diff
@@ -705,32 +682,23 @@ void ColorToFloat(Color color, float* output) {
...
+      "Start/Stop Capture: 'X'\n"
+      "Pan: 'A','D' or \"Left Click + Drag\"\n"
+      "Zoom: 'W', 'S', Scroll or \"Ctrl + Right Click + Drag\"\n"
+      "Vertical Zoom: \"Ctrl + Scroll\"\n"
+      "Select: Left Click\n"
+      "Measure: \"Right Click + Drag\"\n"
+      "Toggle Help: 'H'\n";
+
+  text_renderer_.AddText(help_message, world_x, world_y, GlCanvas::kZValueTextUi,
+                         Color(255, 255, 255, 255), font_size_ * 2);
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitGl/CaptureWindow.cpp:701]: made it clearer by using a constant.

File: OrbitGl/GlCanvas.cpp [hunk 9]
```diff
@@ -84,14 +84,24 @@ GlCanvas::GlCanvas(uint32_t font_size)
...
+}
+
+std::unique_ptr<GlCanvas> GlCanvas::Create(CanvasType canvas_type, uint32_t font_size) {
+  switch (canvas_type) {
+    case CanvasType::kCaptureWindow:
+      return std::make_unique<CaptureWindow>(font_size);
+    case CanvasType::kDebug:
+      return std::make_unique<GlCanvas>(font_size);
+    default:
+      ERROR("Unhandled canvas type (%d)", canvas_type);
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitGl/GlCanvas.cpp:102]: Used UNREACHABLE() instead, which aborts.

File: OrbitGl/ImGuiOrbit.cpp (hunk 1 above)
Comment (pierricgimmig) [reply OrbitGl/ImGuiOrbit.cpp:749]: yeah I removed the function altogether...

File: OrbitGl/ImGuiOrbit.cpp (hunk 2 above)
Comment (pierricgimmig) [reply OrbitGl/ImGuiOrbit.cpp:746]: yeah I removed the function altogether...

File: OrbitGl/TextRenderer.cpp (hunk 3 above)
Comment (pierricgimmig) [reply OrbitGl/TextRenderer.cpp:169]: sure, changed it to a std::array

File: OrbitGl/TimeGraphLayout.cpp (hunk 4 above)
Comment (pierricgimmig) [reply OrbitGl/TimeGraphLayout.cpp:47]: Done.

File: OrbitGl/CaptureWindow.cpp [hunk 10]
```diff
@@ -705,32 +682,23 @@ void ColorToFloat(Color color, float* output) {
...
+  const char* help_message =
+      "Start/Stop Capture: 'X'\n"
+      "Pan: 'A','D' or \"Left Click + Drag\"\n"
+      "Zoom: 'W', 'S', Scroll or \"Ctrl + Right Click + Drag\"\n"
+      "Vertical Zoom: \"Ctrl + Scroll\"\n"
+      "Select: Left Click\n"
+      "Measure: \"Right Click + Drag\"\n"
+      "Toggle Help: 'H'\n";
+
+  text_renderer_.AddText(help_message, world_x, world_y, GlCanvas::kZValueTextUi,
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitGl/CaptureWindow.cpp:700]: Will do shortly.

**Comment by pierricgimmig** [comment]
> I got this crash when I clicked in Misc... Could it be?

Interesting, I can't repro it, neither on Linux nor Windows. Were on on a specific commit or tip of the branch?

**Comment by pierricgimmig** [comment]
@karupayun , @florian-kuebler , I added rounded corner box support:

before (ImGui):

image


after (Custom):

image

I also modified the AddText method so that it returns the exact bounding box of the characters that have been drawn. This can be used to add a perfectly fitting backround box. I the case of the help box, we add a margin to that bounding box, and round the corners.

File: OrbitGl/CaptureWindow.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply OrbitGl/CaptureWindow.cpp:712]: See below.

File: OrbitGl/CaptureWindow.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitGl/CaptureWindow.cpp]: Okay, fine for me.

File: OrbitGl/App.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitGl/App.cpp]: thanks

File: OrbitGl/App.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitGl/App.cpp:500]: thanks!

File: OrbitGl/TimeGraphLayout.cpp (hunk 4 above)
Comment (florian-kuebler) [reply OrbitGl/TimeGraphLayout.cpp]: Thanks!

File: OrbitGl/TextRenderer.cpp (hunk 3 above)
Comment (florian-kuebler) [reply OrbitGl/TextRenderer.cpp]: Okay, thanks!

File: OrbitGl/GlCanvas.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitGl/GlCanvas.cpp]: Okay, thanks!

File: OrbitGl/CaptureWindow.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitGl/CaptureWindow.cpp:712]: Thanks

**Review by florian-kuebler — APPROVED** [review]
LGTM. Thanks!

File: OrbitGl/ImGuiOrbit.cpp (hunk 1 above)
Comment (florian-kuebler) [reply OrbitGl/ImGuiOrbit.cpp]: Okay, thank!

File: OrbitGl/ImGuiOrbit.cpp (hunk 2 above)
Comment (florian-kuebler) [reply OrbitGl/ImGuiOrbit.cpp]: Thanks!

File: OrbitGl/Batcher.h [hunk 11]
```diff
@@ -105,6 +105,13 @@ class Batcher {
   void AddShadedBox(Vec2 pos, Vec2 size, float z, const Color& color,
                     std::shared_ptr<Pickable> pickable,
                     ShadingDirection shading_direction = ShadingDirection::kLeftToRight);
+  void AddRoundedBox(Vec2 pos, Vec2 size, float z, float radius, const Color& color,
```
Comment (karupayun) [inline OrbitGl/Batcher.h:108]: @reichlfl didn't like to add methods to Batcher who includes several primitives. Only a comment.

File: OrbitGl/CaptureWindow.cpp [hunk 12]
```diff
@@ -639,57 +618,50 @@ void CaptureWindow::RenderUI() {
...
+
+      ImGui::EndTabItem();
+    }
 
-    if (first_help_draw_) {
-      // Redraw so that Imgui resizes the
-      // window properly on first draw
-      NeedsRedraw();
-      first_help_draw_ = false;
+    if (ImGui::BeginTabItem("Misc")) {
```
Comment (karupayun) [inline OrbitGl/CaptureWindow.cpp:655]: About this "Misc" tab:

- Why we have the button and present the view as a separate window?
- At least in my case, the pop up window starts with a bigger size than it has to be (a lot of empty space), do you know why could it be?

File: OrbitGl/TextRenderer.h [hunk 13]
```diff
@@ -25,7 +25,8 @@ class TextRenderer {
   void Init();
   void Display(Batcher* batcher);
   void AddText(const char* a_Text, float a_X, float a_Y, float a_Z, const Color& a_Color,
-               uint32_t font_size, float a_MaxSize = -1.f, bool a_RightJustified = false);
+               uint32_t font_size, float a_MaxSize = -1.f, bool a_RightJustified = false,
+               Vec2* out_text_pos = nullptr, Vec2* out_text_size = nullptr);
```
Comment (karupayun) [inline OrbitGl/TextRenderer.h:29]: Nit: I Prefer to have first the out_text arguments and after the RightJustified

**Review by karupayun — APPROVED** [review]
Nice PR Pierric! :)

I like to don't have the debug variables on top of the capture window.
You will have a few issues rebasing this change, sorry for taking too much time.

**Pull request merged**